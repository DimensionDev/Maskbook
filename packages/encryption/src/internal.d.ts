type Readwrite<T> = {
    -readonly [key in keyof T]: T[key]
}
