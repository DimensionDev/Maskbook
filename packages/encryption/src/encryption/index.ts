export * from './Decryption'
export * from './Encryption'
export * from './AppendEncryption'
