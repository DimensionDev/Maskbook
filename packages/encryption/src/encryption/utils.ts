import type { EC_Public_CryptoKey, EC_Private_CryptoKey } from '@masknet/shared-base'
import { EC_KeyCurveEnum } from '../payload'
import type { EncryptIO } from './Encryption'

/** @internal */
export type EphemeralKeys = Map<
    EC_KeyCurveEnum,
    readonly [EC_Public_CryptoKey, EC_Private_CryptoKey] | Promise<readonly [EC_Public_CryptoKey, EC_Private_CryptoKey]>
>
/** @internal */
export function createEphemeralKeysMap(io: Pick<EncryptIO, 'getRandomECKey'>) {
    const ephemeralKeys: EphemeralKeys = new Map()

    // get ephemeral keys, generate one if not found
    async function getEphemeralKey(curve: EC_KeyCurveEnum) {
        if (ephemeralKeys.has(curve)) return ephemeralKeys.get(curve)!
        ephemeralKeys.set(curve, generateEC_KeyPair(io, curve))
        return ephemeralKeys.get(curve)!
    }
    return { ephemeralKeys, getEphemeralKey }
}

/** @internal */
export function fillIV(io: Pick<EncryptIO, 'getRandomValues'>): Uint8Array {
    if (io.getRandomValues) return io.getRandomValues(new Uint8Array(16))
    return crypto.getRandomValues(new Uint8Array(16))
}

async function generateEC_KeyPair(io: Pick<EncryptIO, 'getRandomECKey'>, kind: EC_KeyCurveEnum) {
    if (io.getRandomECKey) return io.getRandomECKey(kind)
    const namedCurve: Record<EC_KeyCurveEnum, string> = {
        [EC_KeyCurveEnum.secp256p1]: 'P-256',
        [EC_KeyCurveEnum.secp256k1]: 'K-256',
    }
    const { privateKey, publicKey } = await crypto.subtle.generateKey(
        { name: 'ECDH', namedCurve: namedCurve[kind] },
        true,
        ['deriveKey'],
    )
    return [publicKey as EC_Public_CryptoKey, privateKey as EC_Private_CryptoKey] as const
}
