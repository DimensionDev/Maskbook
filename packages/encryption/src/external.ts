// This file is the entry for external usage.
export {
    encodePayload,
    parsePayload,
    EC_KeyCurveEnum,
    SocialNetworkEnum,
    SocialNetworkEnumToProfileDomain,
    type EC_Key,
    type Signature,
    type PayloadParseResult,
    type PayloadWellFormed,
    type SupportedPayloadVersions,
} from './payload/index.js'

export {
    decrypt,
    DecryptError,
    DecryptProgressKind,
    DecryptIntermediateProgressKind,
    DecryptErrorReasons,
    type DecryptOptions,
    type DecryptIO,
    type DecryptEphemeralECDH_PostKey,
    type DecryptStaticECDH_PostKey,
    type DecryptProgress,
    type DecryptIntermediateProgress,
    type DecryptReportedInfo,
    type DecryptSuccess,
} from './encryption/index.js'

export { socialNetworkEncoder, socialNetworkDecoder } from './social-network-encode-decode/index.js'

export { type TypedMessage, type Meta } from '../../typed-message/dist/base/index.js'
export * from '../../typed-message/dist/base/core/index.js'
export {
    ProfileIdentifier,
    PostIVIdentifier,
    CheckedError,
    OptionalResult,
    type AESCryptoKey,
    type EC_Public_CryptoKey,
    type EC_Private_CryptoKey,
    type EC_CryptoKey,
} from '@masknet/shared-base'

export * from 'ts-results-es'
