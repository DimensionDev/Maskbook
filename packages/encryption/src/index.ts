export {
    encodePayload,
    parsePayload,
    EC_KeyCurveEnum,
    SocialNetworkEnum,
    SocialNetworkEnumToProfileDomain,
    type EC_Key,
    type Signature,
    type PayloadParseResult,
    type PayloadWellFormed,
    type SupportedPayloadVersions,
} from './payload'

export {
    encrypt,
    EncryptError,
    EncryptErrorReasons,
    type EncryptOptions,
    type EncryptIO,
    type EncryptResult,
    type EncryptTargetE2E,
    type EncryptionResultE2EMap,
    type EncryptTargetPublic,
    type EncryptionResultE2E,
} from './encryption'

export {
    decrypt,
    DecryptError,
    DecryptProgressKind,
    DecryptIntermediateProgressKind,
    DecryptErrorReasons,
    type DecryptOptions,
    type DecryptIO,
    type DecryptEphemeralECDH_PostKey,
    type DecryptStaticECDH_PostKey,
    type DecryptProgress,
    type DecryptIntermediateProgress,
    type DecryptReportedInfo,
    type DecryptSuccess,
} from './encryption'

export {
    //
    type AppendEncryptionIO,
    type AppendEncryptionOptions,
    appendEncryptionTarget,
} from './encryption'

export { socialNetworkEncoder, socialNetworkDecoder, TwitterDecoder } from './social-network-encode-decode'

export {
    type DecodeImageOptions,
    type EncodeImageOptions,
    type ImageTemplateTypes,
    type SteganographyIO,
    steganographyDecodeImage,
    steganographyEncodeImage,
    GrayscaleAlgorithm,
} from './image-steganography'

// TODO: remove them in the future
export { importEC_Key, getEcKeyCurve, importAES as importAESFromJWK } from './utils'
