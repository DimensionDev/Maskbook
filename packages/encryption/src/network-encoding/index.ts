import { EncryptPayloadNetwork } from '../payload/types.js'
import { sharedDecoder, sharedEncoder } from './shared.js'
import { TwitterDecoder, __TwitterEncoder } from './twitter.js'

export function decodeByNetwork(network: EncryptPayloadNetwork, content: string): Array<string | Uint8Array> {
    if (network === EncryptPayloadNetwork.Twitter) {
        return TwitterDecoder(content)
            .map((x) => [x])
            .unwrapOr([])
    }
    const possiblePayload = content.match(/(\u{1F3BC}[\w+/=|]+:\|\|)/giu) || []

    const result: Array<string | Uint8Array> = []
    for (const payload of possiblePayload) {
        const decoded = sharedDecoder(payload)
        if (decoded.isSome()) result.push(decoded.value)
        else result.push(payload)
    }
    return result
}
export function encodeByNetwork(network: EncryptPayloadNetwork, content: string | Uint8Array): string {
    // v38
    if (typeof content === 'string') {
        if (network === EncryptPayloadNetwork.Twitter) return __TwitterEncoder(content)
        return content
    }

    // v37
    if (network === EncryptPayloadNetwork.Twitter) return __TwitterEncoder(content)
    return sharedEncoder(content)
}

export * from './twitter.js'
