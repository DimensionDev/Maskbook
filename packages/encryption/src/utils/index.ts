import { ExceptionKinds, Exception } from '../types'
import { Result } from 'ts-results'
import { encodeArrayBuffer, decodeArrayBuffer, encodeText, decodeText } from '@dimensiondev/kit'

export * from './crypto'

const firstArgString = (e: unknown) => typeof e !== 'string'
const firstArgArrayBuffer = (e: unknown) => e instanceof ArrayBuffer
export const encodeArrayBufferF = wrap(encodeArrayBuffer, firstArgArrayBuffer)
export const decodeArrayBufferF = wrap(decodeArrayBuffer as (x: string) => ArrayBuffer, firstArgString)
export const encodeTextF = wrap(encodeText, firstArgString)
export const decodeTextF = wrap(decodeText, firstArgArrayBuffer)
export const JSONParseF = wrap(JSON.parse, firstArgString)

function wrap<P extends any[], T>(f: (...args: P) => T, valid: (...args: P) => boolean) {
    return <E extends ExceptionKinds>(invalidE: E, throwsE: E) =>
        (...args: P) => {
            const isValid = valid(...args)
            if (!isValid) return new Exception(invalidE, null).toErr()
            return Result.wrap(() => f(...args)).mapErr(Exception.mapErr(throwsE))
        }
}
export async function andThenAsync<T, E, Q, E2>(
    op: Result<T, E> | Promise<Result<T, E>>,
    mapper: (t: T) => Result<Q, E2> | Promise<Result<Q, E2>>,
): Promise<Result<Q, E | E2>> {
    op = await op
    if (op.err) return op
    return mapper(op.val)
}
