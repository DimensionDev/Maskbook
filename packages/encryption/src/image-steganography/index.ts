import { encode, decode, GrayscaleAlgorithm } from '@dimensiondev/stego-js'
import type { EncodeOptions } from '@dimensiondev/stego-js'
import { getDimension } from './utils.js'
import { getPreset, findPreset } from './presets.js'
import { encodeArrayBuffer } from '@dimensiondev/kit'

export { GrayscaleAlgorithm, AlgorithmVersion } from '@dimensiondev/stego-js'

export interface SteganographyIO {
    downloadImage: (url: string) => Promise<ArrayBuffer>
}
export interface EncodeImageOptions extends SteganographyIO {
    data: string | Uint8Array
    password: string
    grayscaleAlgorithm?: GrayscaleAlgorithm
    preset: SteganographyPreset
}
export enum SteganographyPreset {
    Preset2021 = '2021',
    Preset2022 = '2022',
}

export async function steganographyEncodeImage(buf: ArrayBuffer, options: EncodeImageOptions) {
    const { downloadImage, data, password, grayscaleAlgorithm } = options
    const preset = getPreset(options.preset)
    if (!preset) throw new Error('Failed to find preset.')

    const optionalOptions: Partial<Readwrite<EncodeOptions>> = {}
    if (grayscaleAlgorithm) optionalOptions.grayscaleAlgorithm = grayscaleAlgorithm

    const text = typeof data === 'string' ? data : encodeArrayBuffer(data)

    return new Uint8Array(
        await encode(buf, await downloadImage(preset.mask), {
            ...preset.options,
            ...optionalOptions,
            text,
            pass: password,
        }),
    )
}

export type DecodeImageOptions = SteganographyIO & Partial<EncodeOptions> & Pick<EncodeOptions, 'pass'>

export async function steganographyDecodeImage(image: Blob | string, options: DecodeImageOptions) {
    const buffer = typeof image === 'string' ? await options.downloadImage(image) : await image.arrayBuffer()
    const dimension = getDimension(buffer)
    const preset = findPreset(dimension)
    if (!preset) return ''
    return decode(buffer, await options.downloadImage(preset.mask), {
        ...preset.options,
        ...options,
    })
}
