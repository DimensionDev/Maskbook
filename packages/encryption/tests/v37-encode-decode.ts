import { ECDH_K256_Pub } from './setup'
import { test, expect } from '@jest/globals'
import { None, Some } from 'ts-results'
import { encodePayload, AESAlgorithmEnum, parsePayload, PayloadWellFormed } from '../src'
import { importAESFromJWK } from '../src/utils'
import { ProfileIdentifier } from '@masknet/shared-base'

test('Encode v37 payload', async () => {
    const payload: PayloadWellFormed.Payload = {
        version: -37,
        signature: None,
        author: None,
        authorPublicKey: None,
        encrypted: new Uint8Array([0, 1, 2, 3, 4]),
        encryption: {
            type: 'public',
            iv: new Uint8Array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]),
            AESKey: await getAESKey(),
        },
    }
    // test 1
    {
        const encoded = (await encodePayload.NoSign(payload).then((x) => x.unwrap())) as Uint8Array
        expect(encoded).toBeInstanceOf(Uint8Array)

        // These two steps are intended to make sure this Uint8Array is not a partial view to a larger ArrayBuffer.
        expect(encoded.byteOffset).toBe(0)
        expect(encoded.buffer.byteLength).toBe(encoded.length)

        expect(encoded).toMatchSnapshot('Empty payload')

        const parsed = await parsePayload(encoded).then((x) => x.unwrap())
        expect(parsed).toMatchSnapshot('Empty payload parsed')
    }
    // test 2
    {
        const newPayload = { ...payload }
        newPayload.author = Some(new ProfileIdentifier('localhost', 'unknown'))
        const k256Key = await ECDH_K256_Pub()
        newPayload.authorPublicKey = Some(k256Key)
        newPayload.encryption = {
            type: 'E2E',
            iv: payload.encryption.iv,
            ephemeralPublicKey: new Map(),
            ownersAESKeyEncrypted: new Uint8Array([5, 6, 7, 8, 9, 10]),
        }
        newPayload.encryption.ephemeralPublicKey.set(k256Key.algr, k256Key.key)

        const encoded = (await encodePayload.NoSign(newPayload).then((x) => x.unwrap())) as Uint8Array
        expect(encoded).toMatchSnapshot('Full payload')

        const parsed = await parsePayload(encoded).then((x) => x.unwrap())
        expect(parsed).toMatchSnapshot('Full payload parsed')
    }
})

async function getAESKey(): Promise<PayloadWellFormed.PublicEncryption['AESKey']> {
    return {
        algr: AESAlgorithmEnum.A256GCM,
        key: (await importAESFromJWK.AES_GCM_256(testKey)).unwrap(),
    }
}
const testKey = {
    alg: 'A256GCM',
    ext: true,
    /* cspell:disable-next-line */
    k: 'JRhrRKykmnm3SbuNw6OcXF_jiw0gIlW3QiWNV01jeaE',
    key_ops: ['encrypt', 'decrypt'],
    kty: 'oct',
}
