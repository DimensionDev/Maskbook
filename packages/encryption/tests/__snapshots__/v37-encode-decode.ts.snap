// Vitest Snapshot v1

exports[`Encode v37 payload > Empty payload 1`] = `
Uint8Array [
  0,
  151,
  0,
  192,
  192,
  192,
  192,
  147,
  0,
  196,
  32,
  37,
  24,
  107,
  68,
  172,
  164,
  154,
  121,
  183,
  73,
  187,
  141,
  195,
  163,
  156,
  92,
  95,
  227,
  139,
  13,
  32,
  34,
  85,
  183,
  66,
  37,
  141,
  87,
  77,
  99,
  121,
  161,
  196,
  16,
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  196,
  5,
  0,
  1,
  2,
  3,
  4,
]
`;

exports[`Encode v37 payload > Empty payload parsed 1`] = `
{
  "author": OkImpl {
    "err": false,
    "ok": true,
    "val": NoneImpl {
      "none": true,
      "some": false,
    },
  },
  "authorPublicKey": OkImpl {
    "err": false,
    "ok": true,
    "val": NoneImpl {
      "none": true,
      "some": false,
    },
  },
  "encrypted": OkImpl {
    "err": false,
    "ok": true,
    "val": Uint8Array [
      0,
      1,
      2,
      3,
      4,
    ],
  },
  "encryption": OkImpl {
    "err": false,
    "ok": true,
    "val": {
      "AESKey": OkImpl {
        "err": false,
        "ok": true,
        "val": CryptoKey {
          Symbol(kKeyObject): SecretKeyObject {
            Symbol(kKeyType): "secret",
          },
          Symbol(kAlgorithm): {
            "length": 256,
            "name": "AES-GCM",
          },
          Symbol(kExtractable): true,
          Symbol(kKeyUsages): [
            "encrypt",
            "decrypt",
          ],
        },
      },
      "iv": OkImpl {
        "err": false,
        "ok": true,
        "val": Uint8Array [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
          12,
          13,
          14,
          15,
        ],
      },
      "type": "public",
    },
  },
  "signature": OkImpl {
    "err": false,
    "ok": true,
    "val": NoneImpl {
      "none": true,
      "some": false,
    },
  },
  "version": -37,
}
`;

exports[`Encode v37 payload > Full payload 1`] = `
Uint8Array [
  0,
  151,
  0,
  169,
  108,
  111,
  99,
  97,
  108,
  104,
  111,
  115,
  116,
  167,
  117,
  110,
  107,
  110,
  111,
  119,
  110,
  2,
  196,
  33,
  2,
  175,
  219,
  85,
  96,
  10,
  190,
  135,
  73,
  185,
  68,
  70,
  147,
  119,
  167,
  140,
  77,
  96,
  82,
  43,
  182,
  72,
  66,
  204,
  240,
  130,
  10,
  34,
  83,
  70,
  168,
  229,
  140,
  148,
  1,
  196,
  6,
  5,
  6,
  7,
  8,
  9,
  10,
  196,
  16,
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  129,
  161,
  50,
  196,
  33,
  2,
  175,
  219,
  85,
  96,
  10,
  190,
  135,
  73,
  185,
  68,
  70,
  147,
  119,
  167,
  140,
  77,
  96,
  82,
  43,
  182,
  72,
  66,
  204,
  240,
  130,
  10,
  34,
  83,
  70,
  168,
  229,
  140,
  196,
  5,
  0,
  1,
  2,
  3,
  4,
]
`;

exports[`Encode v37 payload > Full payload parsed 1`] = `
{
  "author": OkImpl {
    "err": false,
    "ok": true,
    "val": SomeImpl {
      "none": false,
      "some": true,
      "val": ProfileIdentifier {
        "network": "localhost",
        "userId": "unknown",
      },
    },
  },
  "authorPublicKey": OkImpl {
    "err": false,
    "ok": true,
    "val": SomeImpl {
      "none": false,
      "some": true,
      "val": {
        "algr": 2,
        "key": CryptoKey {},
      },
    },
  },
  "encrypted": OkImpl {
    "err": false,
    "ok": true,
    "val": Uint8Array [
      0,
      1,
      2,
      3,
      4,
    ],
  },
  "encryption": OkImpl {
    "err": false,
    "ok": true,
    "val": {
      "ephemeralPublicKey": {
        "2": OkImpl {
          "err": false,
          "ok": true,
          "val": {
            "algr": 2,
            "key": CryptoKey {},
          },
        },
      },
      "iv": OkImpl {
        "err": false,
        "ok": true,
        "val": Uint8Array [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
          12,
          13,
          14,
          15,
        ],
      },
      "ownersAESKeyEncrypted": OkImpl {
        "err": false,
        "ok": true,
        "val": Uint8Array [
          5,
          6,
          7,
          8,
          9,
          10,
        ],
      },
      "type": "E2E",
    },
  },
  "signature": OkImpl {
    "err": false,
    "ok": true,
    "val": NoneImpl {
      "none": true,
      "some": false,
    },
  },
  "version": -37,
}
`;

exports[`Encode v37 payload: Empty payload 1`] = `Uint8Array [ 009700c0c0c0c09300c42025186b44aca49a79b749bb8dc3a39c5c5fe38b0d202255b742258d574d6379a1c410000102030405060708090a0b0c0d0e0fc4050001020304 ]`;

exports[`Encode v37 payload: Empty payload parsed 1`] = `
Object {
  "author": Ok(None),
  "authorPublicKey": Ok(None),
  "encrypted": Ok(Uint8Array [ 0001020304 ]),
  "encryption": Ok(Object {
    "AESKey": Ok(CryptoKey { [opaque crypto key material] }),
    "iv": Ok(Uint8Array [ 000102030405060708090a0b0c0d0e0f ]),
    "type": "public",
  }),
  "signature": Ok(None),
  "version": -37,
}
`;

exports[`Encode v37 payload: Full payload 1`] = `Uint8Array [ 009700a96c6f63616c686f7374a7756e6b6e6f776e02c42102afdb55600abe8749b944469377a78c4d60522bb64842ccf0820a225346a8e58c9401c40605060708090ac410000102030405060708090a0b0c0d0e0f81a132c42102afdb55600abe8749b944469377a78c4d60522bb64842ccf0820a225346a8e58cc4050001020304 ]`;

exports[`Encode v37 payload: Full payload parsed 1`] = `
Object {
  "author": Ok(Some(ProfileIdentifier {
    "network": "localhost",
    "userId": "unknown",
  })),
  "authorPublicKey": Ok(Some(Object {
    "algr": 2,
    "key": CryptoKey { [opaque crypto key material] },
  })),
  "encrypted": Ok(Uint8Array [ 0001020304 ]),
  "encryption": Ok(Object {
    "ephemeralPublicKey": Object {
      "2": Ok(Object {
        "algr": 2,
        "key": CryptoKey { [opaque crypto key material] },
      }),
    },
    "iv": Ok(Uint8Array [ 000102030405060708090a0b0c0d0e0f ]),
    "ownersAESKeyEncrypted": Ok(Uint8Array [ 05060708090a ]),
    "type": "E2E",
  }),
  "signature": Ok(None),
  "version": -37,
}
`;
